[Billy Ho Cheng En] iP

billyhoce opened the pull request:
# :smiling_imp: ***MEAN***-**DUKE** :smiling_imp:

>"In the crucible of adversity, a touch of assertiveness can be the catalyst for self-discovery and resilience, forging a stronger version of oneself." – [ChatGPT](https://chat.openai.com/share/1dd911ae-addf-4a3f-a77c-069032d82702)

MeanDuke constantly pushes you to the edge of ***insanity*** as it questions your every decision while keeping track of things you need to do.

To achieve the most out of this chatbot, you need the following:

- [ ] Java 11 or above installed on your computer
- [ ] Tasks to be completed
- [ ] A strong mental fortitude

All you need to do is,

1. download `MeanDuke.jar` from [here](https://upload.wikimedia.org/wikipedia/en/9/9a/Trollface_non-free.png).
2. double-click it.
3. add your tasks.
4. Prepare to be verbally berated

Here is just a snippet of what is to come:
```java
    public InvalidCommandException() {
        super();
        this.errorMessage = "What are you saying? Read the user manual, it was written for a reason";
    }
```

Disclaimer: The authors of this chatbot is in no way, shape, or form responsible for any of the following:
-  Mental trauma caused by interaction with the chatbot
-  Reputational damage incurred by use of this chatbot for official businesses
-  Property damage caused by the user after losing their mind over this chatbot's snark replies

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

billyhoce committed:
Renamed, edited intro message, added outro message

billyhoce committed:
Added scanner to take user input and ability to echo back messages

billyhoce committed:
Added task list class and ability to create and add to a list

billyhoce committed:
Added new Task class, added ability to mark and unmark tasks as Done

billyhoce committed:
Added ToDo, Deadline, and Event categories for tasks

billyhoce committed:
Removed special characters from logo and spacer. Added automated Text UI Testing

billyhoce committed:
Added handling of incorrect inputs and correct usage using exceptions

billyhoce committed:
Added "Delete" functionality

billyhoce committed:
Updated code to follow Java coding standard closer

billyhoce committed:
Update code to follow Java Code Conventions

billyhoce committed:
Add support for loading a save file containing the state of a TaskList

billyhoce committed:
Add automated saving to disk when task list changes. Update JavaDoc comments.

billyhoce committed:
Merge branch 'branch-Level-7'

billyhoce committed:
Add proper handling of Dates and Times for Deadline and Event classes using java.time

billyhoce committed:
Merge branch 'branch-Level-8'

billyhoce committed:
Reorganise code using more OOP. Extract into classes UI, Command, Parser

Most of the logic from the MeanDuke is extracted out into other classes. The Storage class has yet to be implemented and hence this version does not support storage functionality.

billyhoce committed:
Fix implementation of Storage class and modify spacer length

billyhoce committed:
Merge branch 'branch-A-MoreOOP'

billyhoce committed:
Package source files

billyhoce committed:
Merge remote-tracking branch 'origin/add-gradle-support'

billyhoce committed:
Add Gradle implementation, including Checkstyle

billyhoce committed:
Add some JUnit tests for Parser and Task classes, works with Gradle

billyhoce committed:
Merge branch 'branch-A-JUnit'

billyhoce committed:
Change name of executable .jar file to MeanDuke.jar

billyhoce committed:
Add JavaDocs for public methods. Change Parser methods to Protected

billyhoce committed:
Comply with Java Coding Standard

billyhoce committed:
Add find functionality to filter through TaskList

billyhoce committed:
Add find feature to filter TaskList

billyhoce committed:
Merge branch 'branch-Level-9'

billyhoce committed:
Merge branch 'branch-A-CodingStandard'

billyhoce committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/core/Parser.java

billyhoce committed:
Fix switch-case formatting

billyhoce committed:
Fix merging bugs

billyhoce committed:
Fix style using Checkstyle

bennyLCK commented on src/main/java/duke/core/Ui.java:
```diff
@@ -0,0 +1,77 @@
+            + "$$ $$ $$/$$ |$$$$$/    $$$$$$$$ |$$ $$ $$ |$$$$$$/ $$ |  $$ |$$ |  $$ |$$$$$  \\  $$$$$/\n"
+            + "$$ |$$$/ $$ |$$ |_____ $$ |  $$ |$$ |$$$$ |        $$ |__$$ |$$ \\__$$ |$$ |$$  \\ $$ |_____\n"
+            + "$$ | $/  $$ |$$       |$$ |  $$ |$$ | $$$ |        $$    $$/ $$    $$/ $$ | $$  |$$       |\n"
+            + "$$/      $$/ $$$$$$$$/ $$/   $$/ $$/   $$/         $$$$$$$/   $$$$$$/  $$/   $$/ $$$$$$$$/\n";
```
***WOO HOOOOOO!!!*** What a cool LOGO!

bennyLCK commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,24 @@
+ */
+public abstract class AddCommand extends Command {
+
+    private final TaskList taskList;
```
Perhaps you can also consider referring to the `TaskList` objects as arguments to method calls instead of keeping a reference for each `AddCommand` instance or its subtype, which is introduced in the guide by providing it as an argument to the `execute` method.

bennyLCK commented on src/main/java/duke/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,49 @@
+                false, this.deadlineDate, this.deadlineTime));
+        Ui.printMessage("Added Deadline task: " + description
+                + "(by: " + deadlineDate
+                + (deadlineTime == null ? "" : " " + deadlineTime)
```
Nit: The use of `this.` reference to `deadlineDate` and `deadlineTime` was not consistent here. We all know how hard it is to constantly use the `this` keyword, so just do your best and try to use it for all references in the same method.

bennyLCK commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,31 @@
+     */
+    public static String getUsage() {
+        return "Usage:";
+    }
```
I really like your initiative in providing functionalities to help the user abide by proper command formats through this method.

bennyLCK commented on src/main/java/duke/commands/AddEventCommand.java:
```diff
@@ -0,0 +1,56 @@
+        super.getTaskList().add(new Event(description, false, fromDate, fromTime, toDate, toTime));
+        Ui.printMessage("Added Event task: " + description
+                + "\n(from: " + fromDate + (fromTime == null ? "" : " " + fromTime)
+                + "\n   to: " + toDate + (toTime == null ? "" : " " + toTime) + ")");
```
Perhaps you can consider using the `String.format` method provided by the String class to concatenate the components together?

bennyLCK commented on src/main/java/duke/commands/InvalidCommandException.java:
Perhaps it would be best if this class file was put into the exceptions package?

bennyLCK commented on src/main/java/duke/core/Parser.java:
```diff
@@ -0,0 +1,188 @@
+            } catch (IndexOutOfBoundsException e) {
+                throw new InvalidCommandException(AddCommand.getUsage());
+            }
+        case "LIST":
+            cmd = new ListCommand(taskList);
+            break;
+        case "MARK":
+            try {
+                cmd = parseMark(commandAndRemaining[1].strip(), taskList);
+                break;
+            } catch (IndexOutOfBoundsException | NumberFormatException e) {
+                throw new InvalidCommandException(MarkCommand.getUsage());
+            }
+        case "UNMARK":
+            try {
+                cmd = parseUnmark(commandAndRemaining[1].strip(), taskList);
+                break;
+            } catch (IndexOutOfBoundsException | NumberFormatException e) {
+                throw new InvalidCommandException(UnmarkCommand.getUsage());
+            }
+        case "DELETE":
+            try {
+                cmd = parseDelete(commandAndRemaining[1].strip(), taskList);
+                break;
+            } catch (IndexOutOfBoundsException | NumberFormatException e) {
+                throw new InvalidCommandException(DeleteCommand.getUsage());
+            }
+        case "END":
+            cmd = new ExitCommand();
+            break;
+        case "FIND":
+            try {
+                cmd = new FindCommand(taskList, commandAndRemaining[1]);
+            } catch (IndexOutOfBoundsException e) {
+                throw new InvalidCommandException(FindCommand.getUsage());
+            }
+            break;
+        default:
+            throw new InvalidCommandException();
+        }
```
I am very glad you have ensured the non-indentation of switch-case statements which is not the case for many since it is a bit strange to us. This is a clear sign that you have done your due diligence in checkstyling your code which is reflected everywhere else too! Good job!

bennyLCK commented on src/main/java/duke/core/Parser.java:
```diff
@@ -0,0 +1,188 @@
+     * @return The AddCommand which adds the new Task to the TaskList.
+     * @throws InvalidCommandException if the user input is not a valid AddCommand.
+     */
+    protected static AddCommand parseAdd(String inputWithoutAdd, TaskList taskList) throws InvalidCommandException {
```
I really like your use of `protected` access modifier to only allow classes in same package which is all that is needed to access such methods in the `Parser` class.

bennyLCK commented on src/main/java/duke/core/Storage.java:
```diff
@@ -0,0 +1,105 @@
+     * @return true or false depending on the string
+     * @throws java.util.InputMismatchException if string is not a valid boolean value
+     */
+    private static boolean parseSaveBoolean(String str) throws java.util.InputMismatchException {
+        if (str.equals("true")) {
+            return true;
+        } else if (str.equals("false")) {
+            return false;
+        } else {
+            throw new java.util.InputMismatchException();
+        }
+    }
```
I like how you accounted for corrupted file scenario which is a stretch goal. Continue hustling!

bennyLCK commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,45 @@
+        this.deadlineTime = deadlineTime;
+    }
+
+    @Override
+    public String saveString() {
+        return "DEADLINE" + "\n"
+                + super.saveString() + "\n"
+                + this.deadlineDate
+                + (this.deadlineTime == null ? "" : ";" + this.deadlineTime);
+    }
```
Perhaps you can consider alternative ways of performing String concatenation as the `+` operator in Java is a very inefficient way of performing it. I would recommend using the `String.format()` method provided by the String class or using a `StringBuilder` instance to store intermediate states of a String message.

bennyLCK reviewed (commented):
Overall, it LGTM. I am really amazed by the level of organization in your codes and the handling of cases in a clear and concise manner. Some Nits I have is where you sometimes lack consistency in using the `this` keyword, the `InvalidCommandException` class being located in the `commands` package instead of the `exceptions` package and using `+` operators in many printing methods instead of perhaps using String.format().

Jajared commented on src/main/java/duke/commands/InvalidCommandException.java:
```diff
@@ -0,0 +1,34 @@
+    public InvalidCommandException() {
+        super();
+        this.errorMessage = "What are you saying? Read the user manual, it was written for a reason";
+    }
```
Great use of method overloading here to manage constructor calls with different parameters. While not an enforced coding standard, do consider placing the constructor with no arguments (default constructor) above the constructors with parameters.

Jajared commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,31 @@
+     *
+     * @throws MeanDukeException if execution issues occur.
+     */
+    public abstract void execute() throws MeanDukeException;
```
Good use of abstract methods to create blueprints for your classes/interfaces!

Jajared commented on src/main/java/duke/core/MeanDuke.java:
```diff
@@ -0,0 +1,44 @@
+            String userInput = Ui.readInput();
+            Ui.printSpacer();
+            try {
+                Command cmd = Parser.parseUserInput(userInput, taskList);
```
This is a great way to abstract out the parsing of user input! However, do consider using a more specific variable name such as `userCommand` to improve readability

Jajared commented on src/main/java/duke/core/Parser.java:
```diff
@@ -0,0 +1,188 @@
+        String[] commandAndRemaining = userInput.strip().split(" ", 2);
+        String commandUpper = commandAndRemaining[0].toUpperCase();
+        Command cmd;
+        switch (commandUpper) {
```
Good use of abstraction to parse the individual commands instead of lumping all of the logic in a code block! This improves readability and can help in tracing potential bugs by retracing the methods!

Jajared commented on src/main/java/duke/core/Storage.java:
```diff
@@ -0,0 +1,105 @@
+     * @return true or false depending on the string
+     * @throws java.util.InputMismatchException if string is not a valid boolean value
+     */
+    private static boolean parseSaveBoolean(String str) throws java.util.InputMismatchException {
```
Good effort in managing exceptions! However, do consider using InputMismatchException and importing the required package at the top instead of specifying the full package name java.util.InputMismatchException? This is to align with the other imports and usage of packages.

Jajared commented on src/test/java/duke/core/ParserTest.java:
```diff
@@ -0,0 +1,64 @@
+import duke.tasks.TaskList;
+
+
+public class ParserTest {
```
Very detailed test cases provided! Do consider adding javadocs to your code to improve readability 😄

Jajared reviewed (commented):
Overall, good and clean code with well thought out logic! Coding standards and quality are generally well adhered to. Keep up the good work!

billyhoce committed:
Add GUI from JavaFX Tutorial

billyhoce committed:
Convert MeanDuke to a fully GUI

billyhoce committed:
Remove UI class,  I/O redirection testing, reorganise exception classes

billyhoce committed:
Merge branch 'branch-Level-10'

billyhoce committed:
Update gradle point of entry to Launcher for GUI

billyhoce committed:
Add assertions to document important assumptions

Assertions can be used to define assumptions about the program state. A
failure indicates a possible bug in the code.

Let's add assertions to achieve the following:
* Ensure that a reply from Mean Duke cannot be empty
* Ensure that tasks with no description cannot be created
* Ensure that when saving a task, its data cannot be empty

billyhoce committed:
Improve Code Quality

Code quaity guidelines like Maximizing readability and following
standards are important to allow others to understand and build onto
our code.

Let's take these steps to improve code quality:
* Follow SLAP by extracting methods from long and deep methods
* Being consistant in the use of the 'this.' keyword
* Removing magic numbers by adding them as constants
* Reordering method positions to follow standards

billyhoce committed:
Merge pull request #1 from billyhoce/branch-A-Assertions

Add assertions to document important assumptions

Assertions can be used to define assumptions about the program state. A failure indicates a possible bug in the code.

Let's add assertions to achieve the following:

* Ensure that a reply from Mean Duke cannot be empty
* Ensure that tasks with no description cannot be created
* Ensure that when saving a task, its data cannot be empty

billyhoce committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/core/MainWindow.java

billyhoce committed:
Merge pull request #2 from billyhoce/branch-A-CodeQuality

Improve Code Quality

Code quality guidelines like Maximizing readability and following standards are important to allow others to understand and build on our code.

Let's take these steps to improve code quality:

* Follow SLAP by extracting methods from long and deep methods
* Being consistent in the use of the 'this.' keyword
* Removing magic numbers by adding them as constants
* Reordering method positions to follow standard

billyhoce committed:
Use GitHub Actions to set up Continuous Integration

billyhoce committed:
Fix styling issues brought up by checkstyle

billyhoce committed:
Add friendlier syntax, partial keywords

billyhoce committed:
Fix styling issue in enum classes

billyhoce committed:
Merge branch 'master' into branch-A-BetterGui

billyhoce committed:
Update the GUI with a theme

billyhoce committed:
Merge branch 'branch-A-BetterGui'

billyhoce committed:
Rename duke package into meanduke package

billyhoce committed:
Make friendly syntax more friendly

Previously, the parser looks for a command that contains user input as
a substring. Resulting in situations like "d" becoming "todo", when it
should become "deadline".

With this change, the command that starts with the user input is chosen

billyhoce committed:
Add User Guide and Product Screenshot

billyhoce committed:
Crop pixel line at top of product screenshot

billyhoce committed:
Add message when list called on empty tasklist

billyhoce committed:
Update README.md
